### ISSUE TITLE
Possibility to execute test cases

### ISSUE THREAD

**Issue by robotframework-ride** [robotframework/RIDE#91]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57) by jrvilda on 18 Nov 2008*

**Explain the new feature or other enhancement idea you have.**

Add the opcion &lt;executed&gt; in tools´s menu.
The opcion &lt;executed&gt;  launch pybot.py with command :

pybot.bat --test &lt;name testcase select&gt; Path

another option would be to open a window to click on the option &lt;executed&gt;
the tools menu, where it appears that the command was launched, so it is
saved and there were always able to customize it with a more consistent
&lt;NametestCase&gt;

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c1) by @yanne on 20 Nov 2008*


This has been discussed before, and it may be implemented somewhere in the future.

However, the proper solution is somewhat complicated since we need to be able to
support all Robot Framework command line options and also show the reports and logs
automatically. Additionally, it is really easy to have a command prompt open when
editing the test cases and just run the tests from the command line.

All in all, this is not going to be implemented in short term.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c2) by @pekkaklarck on 4 Dec 2008*


We also needed to support both Python and Jython. This won't be done for 1.0 version.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c3) by vesa.kur... on 5 Dec 2008*


What about freely configurable "button".
So user could create several button with names and inserting what ever script 
execution for the button.

RIDE would open either cmd /terminal window and execute the script written to button.

This would enable e.g. creating SVN update button, Runtest button etc..

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c4) by bryan.oakley on 31 Mar 2009*


I agree with the previous comment that it would be best to support configurable
buttons. For example, it would be nice if I could edit ~/.robotiderc and add
something like:

# define a function named "run"
def run ():
    # whatever python code I want

# add a button with the label "Run" that calls the function "run"
add_button("Run",run)

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c5) by @pekkaklarck on 6 May 2009*


<a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=135" title="support running tests from within RIDE"> Issue 135 </a> has been merged into this issue.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c6) by @pekkaklarck on 25 Jun 2009*


<a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=177" title="plugin for launching robot tests"> Issue 177 </a> has been merged into this issue.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c7) by @pekkaklarck on 25 Jun 2009*


This will be implemented as a plugin. Hopefully sooner than later.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c8) by jrvilda on 8 Jul 2009*


	
I use this plugin.
It is not very good.
But it may be an idea

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c9) by jrvilda on 10 Sep 2009*


Es mejorable, pero con pocos cambios puede ser util.
pulsar el boton actualizar, seleccionar y ejecutar.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c10) by jrvilda on 10 Sep 2009*


Solo para windows

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c12) by @yanne on 18 Nov 2009*


The focus of 0.20 is solely in the plugin API, everything else is scoped to
the next release

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c13) by @yanne on 11 Jan 2010*


We have made quite a bit of progress with generic run plugin (<a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=253" title="Generic plugin for running any command in the system"> issue 253 </a>). That plugin 
also allows for executing tests and storing execution configurations. Since 
implementing easily configurable support for all RF command line options is a big 
task and the benefit is not that significant, I don't think this specific test 
execution plugin is going to be implemented in short term.

Anyone interested in executing tests, please take a look at the run plugin in current 
head of the repository and offer any feedback in <a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=253" title="Generic plugin for running any command in the system"> issue 253 </a>.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c14) by @pekkaklarck on 11 Jan 2010*


One of the benefits a "real" test execution plugin would have is that it could show
how the execution proceeds using the listener API of the core framework. It could
also pause the test execution and, if the listener API is enhanced, stop it for good.
These are great benefits and I'd like to see them in RIDE 1.0 but for now the generic
running possibilities provided by <a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=252" title="Implement default test suite, test case, and keyword editors as plugin"> issue 252 </a> ought to be enough.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c17) by @pekkaklarck on 2 Feb 2010*


<a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=365" title="Could RIDE launch the suite execution directly, instead of pybot or jybot?"> Issue 365 </a> has been merged into this issue.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c18) by @pekkaklarck on 24 Jul 2010*


<a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=481" title="Run Pybot"> Issue 481 </a> has been merged into this issue.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c21) by @jrantanen on 20 Jan 2011*


Bryan provided the TestRunner plugin and we started to test it.

Following things needs to be done before this issue can be closed: 
1) Change settings to use the Plugin API settings
2) Enable/Disable "Start Robot" and "Stop a running test" on menubar item, toolbar button, and shortcut. API available as explained in <a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=679" title="Allow disabling and enabling menu items, toolbar buttons and shortcuts from plugin API"> issue 679 </a>.
3) Test and document that run scripts can be added to settings file.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c22) by @mkorpela on 21 Jan 2011*


Tested in Ubuntu 10.04 LTS 64 bit.

Some issues about the run plugin:
1) I can't see any way to reopen the run panel without running tests

2) Doesn't remember my execution profile (and it isn't selected by default)

3) TestSuiteTree is broken after I have closed the run panel and try to run tests
&gt;&gt;
Traceback (most recent call last):
  File "/usr/lib64/python2.6/dist-packages/wx-2.8-gtk2-unicode/wx/_core.py", line 14614, in &lt;lambda&gt;
    lambda event: event.callable(*event.args, **event.kw) )
  File "./src/robotide/contrib/testrunner/testrunnerplugin.py", line 878, in _post_result
    self._tree.SetState(longname, "fail")
  File "./src/robotide/contrib/testrunner/TestSuiteTreeCtrl.py", line 219, in SetState
    node = self._nodes[testId]
KeyError: u'Simple Testsuite With Different Namespaces'
&lt;&lt;

4) Plugin tree doesn't recognize rename or delete of a test case
5) Button in Plugin tree is transparent
6) Plugin tree doesn't recognize a new suite

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c23) by bryan.oakley on 21 Jan 2011*


What do you mean by "reopen the run panel"? Are you saying you delete
the notebook tab and there's no way to get it back?

Simple solution: don't delete the notebook tab :-) But you're right,
there's no way to get it back. It's interesting - we've been using
this plugin for a year or so and not once has somebody mentioned this.

That's a bug that I can't reproduce with the current version. Do you
get any stack traces on stdout?

I can confirm that.

I don't think RIDE publishes those events so there's no way for the
plugin to know they happened. The tree in the plugin refreshes when
you reselect that tab so I *think* it's only a problem if you have the
run tab visible when you make a rename. The normal case is that you
have the edit window open, you do a rename, and when you select the
run tab it refreshes.

Not sure what you mean by that. It should appear as just a triangle
(not a square button). Are you saying the triangle is transparent? (if
so, how can you see it at all?)

Again, I don't think RIDE publishes that event so the runner has no
way of knowing a suite was added.

You should create bug reports for each of these issues so they can be tracked.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c24) by bryan.oakley on 21 Jan 2011*


(yikes! I responded in-line via email and I see that somewhere along the line all the context was lost)

Since it's not obvious, the "that's a bug I can't confirm" statement refers to saving the selected profile. As the saying goes "it works for me". 

The comments about RIDE not publishing events has to do with suites being added, and test cases being renamed or deleted.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c25) by bryan.oakley on 21 Jan 2011*


I committed a change that I think fixes <a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=3" title="Find unused User Keywords"> issue #3 </a> mentioned above.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c26) by @mkorpela on 21 Jan 2011*


RIDE publishes the following messages (I don't remember when they were added .. possibly just recently):
RideItemNameChanged (for user keyword or test case name change)
RideTestCaseRemoved

.. and I think your right about there being no message of new suite -- but we can add it before making the release.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c27) by bryan.oakley on 21 Jan 2011*


I pushed a change to redraw the tree on RideItemNameChanged and RideTestCaseRemoved, which I think solves <a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=4" title="Tag refactoring"> issue #4 </a>, above.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c28) by @jrantanen on 25 Jan 2011*


I have now done following refactoring and bug fixes:
1) Plugin API is used for storing and accessing settings, <a href="https://code.google.com/p/robotframework-ride/source/detail?r=42897f0a7a">revision 42897f0a7a</a>.
2) Plugin API is used for creating and enabling/disabling menu items, toolbar buttons and shortcuts, <a href="https://code.google.com/p/robotframework-ride/source/detail?r=d93378a58a">revision d93378a58a</a>
3) If there is unsaved changes, saves all the files if autosave is on, otherwise asks user input, <a href="https://code.google.com/p/robotframework-ride/source/detail?r=3a0c2f967d">revision 3a0c2f967d</a>.
4) Some other minor bug fixes

Following things are still needed:
1) Check the process stopping (see FIXME from the code)
2) Add documentation how to add own scripts to run selection.


I think that there are a bunch of configuration possibilities that could be added, but those can be implemented in forthcoming releases. We should get user feedback now and add functionality based on their needs.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c29) by @mkorpela on 26 Jan 2011*


Process stopping (in Ubuntu 10.04 64 bit):
I have no idea why this works but the process pid is off by two!?!?!
if I replace asyncproc:kill[198] with os.kill(self.pid()+2, signal) everything just works.

also self.__process.kill(signal) will try to kill wrong process (added in Python 2.6).

for some reason self.__process.terminate() works (added in Python 2.6).

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c30) by @mkorpela on 26 Jan 2011*


Windows 7 64bit (using it as a normal user -- not admin):
er\asyncproc.py", line 198, in kill
    os.kill(self.pid(), signal)
WindowsError: [Error 5] Access is denied

conclusion: You can't use run plugin if you don't have admin rights.

There was a numberof other errors that happen after this but I think this is the root cause.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c31) by @mkorpela on 26 Jan 2011*


Windows 7 64bit (using it as an admin):
pybot output is now presented on time (it wasn't in "normal" user mode).

Still can't seem to find a way to really stop the process (self.__process.temintat() didn't work).

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c32) by bryan.oakley on 26 Jan 2011*


Perhaps if the robot listener added a method to send its actual pid to the remote listener we could work around some of these problems? I'm guessing the ubuntu problem might be that the process is actually a script that calls python that calls pybot, or something like that so that the pid of the process that gets exec'd isn't the actual pid of the pybot process. Though, it could be a bug with asyncproc, too. Unfortunately I don't have access to a ubuntu system to test.

Here's a solution that might be worth considering: change the SocketListener to use XMLRPC to talk to the plugin. This allows two way communication, since the listener gets back the result of each remote command. Then, the plugin could return "QUIT" if the user presses the stop button. If the listener gets that it could cause the test running mechanism to stop since it's in the same process. Of course, this won't help if the test runner is hung, but it would work under normal circumstances.

I actually have a lot of this code working in my workbench project so I know it's possible. It's a bigger project than just hacking around in asyncproc but it might be the better long term solution.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c33) by @jrantanen on 26 Jan 2011*


I now added custom profile which allows user to write or browse with file browser the script to be used. See <a href="https://code.google.com/p/robotframework-ride/source/detail?r=bd229f9e92">revision bd229f9e92</a>. I will also update documentation.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c34) by @mkorpela on 26 Jan 2011*


I fixed the windows termination problem. It doesn't terminate nicely (produce robot output) but it kills the process.

Unix one can be done more elegantly.

We could send the SIGINT to all child processes of the spawned process (and all of the children of those etc.) but this seems to require that we call ps and get the process parent id from the output.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c35) by bryan.oakley on 26 Jan 2011*


Instead of doing a ps, why not just have the listener send it's pid to the plugin? There's no reason that SocketListener couldn't send this information across the socket. I think that's the most elegant solution.

For example, every start_suite message could add the result of os.pid() to the payload before sending it across the wire.

Wouldn't that work?

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c36) by @mkorpela on 27 Jan 2011*


@&#8288;bryan: Your solution for the pid problem works.

**Comment by robotframework-ride** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=57#c38) by bryan.oakley on 27 Jan 2011*


I dont know why I didn't think of that solution to the pid problem before. Glad to hear it works.
